odoo/addons/base/models/ir_qweb.py

```diff
@@ -148,31 +722,880 @@ def is_child_view(view_name):
+                        yield f' {{str(escape(str(name)))}}="{{str(escape(str(value)))}}"'
+                attrs = None
+        """).strip(), indent))
+
+        # close the open tag
+        if 't-tag-close' in el.attrib:
+            self._append_text('>', options)
+        else:
+            self._append_text('/>', options)
+
+        return code
+
+    def _compile_directive_tag_close(self, el, options, indent):
+        """ Compile the closing tag of the given element into string.
+        Returns an empty list because it's use only `_append_text`.
+        """
+        el_tag = el.attrib.pop("t-tag-close", None)
+        if el_tag:
+            self._append_text(f'</{el_tag}>', options)
+        return []
+
+    def _compile_directive_set(self, el, options, indent):
+        """Compile `t-set` expressions into a python code as a list of
+        strings.
+
+        There are 3 kinds of `t-set`:
+        * `t-value` containing python code;
+        * `t-valuef` containing strings to format;
+        * whose value is the content of the tag (being Markup safe).
+
+        The code will contain the assignment of the dynamically generated value.
+        """
+        varname = el.attrib.pop('t-set')
+        code = self._flush_text(options, indent)
+
+        if 't-value' in el.attrib:
+            if varname == T_CALL_SLOT:
+                raise ValueError('t-set="0" should not contains t-value or t-valuef')
+            expr = el.attrib.pop('t-value') or 'None'
+            expr = self._compile_expr(expr)
+
+            # The content is considered empty.
+            el.attrib.pop('t-inner-content')
+        elif 't-valuef' in el.attrib:
+            if varname == T_CALL_SLOT:
+                raise ValueError('t-set="0" should not contains t-value or t-valuef')
+            exprf = el.attrib.pop('t-valuef')
+            expr = self._compile_format(exprf)
+
+            # The content is considered empty.
+            el.attrib.pop('t-inner-content')
+        else:
+            # set the content as value
+            def_name = self._make_name("qweb_t_set")
+            content = self._compile_directive(el, options, 'inner-content', indent + 1) + self._flush_text(options, indent + 1)
+            if content:
+                code.append(self._indent(f"def {def_name}():", indent))
+                code.append(self._indent("attrs = None", indent + 1))
+                code.extend(content)
+                expr = f"Markup(''.join({def_name}()))"
+            else:
+                expr = "''"
+
+        code.append(self._indent(f"values[{repr(varname)}] = {expr}", indent))
+        return code
+
+    def _compile_directive_value(self, el, options, indent):
+        """Compile `t-value` expressions into a python code as a list of strings.
+
+        This method only check if this attributes is on the same node of a
+         `t-set` attribute.
+        """
+        if not options.get('t-set') is None:
```

Comment:
Seems cleaner (or I make a mistake and options can have 't-set' equals to None ?):
`if 't-set' not in options:`